src/pool/srv_util.c

```diff
@@ -529,3 +531,257 @@ int ds_pool_get_failed_tgt_idx(const uuid_t pool_uuid, int **failed_tgts,
+		return 1;
+	else
+		return (*pl_rank == -1) ? -DER_NOSYS : 2;
+}
+
+struct exclude_targets_arg {
+	uuid_t		 eta_pool_id;
+	d_rank_t	 eta_pl_rank;
+	d_rank_t	*eta_ranks;
+	int		*eta_tgts;
+	int		 eta_nr;
+};
+
+static void
+free_exclude_targets_arg(struct exclude_targets_arg *eta)
+{
+	D_ASSERT(eta != NULL);
+	if (eta->eta_ranks != NULL)
+		D_FREE(eta->eta_ranks);
+	if (eta->eta_tgts != NULL)
+		D_FREE(eta->eta_tgts);
+	D_FREE(eta);
+}
+
+static struct exclude_targets_arg *
+alloc_exclude_targets_arg(uuid_t pool_id, int *tgt_ids, int tgt_cnt,
+			  d_rank_t pl_rank)
+{
+	struct exclude_targets_arg	*eta;
+	d_rank_t			 rank;
+	int				 i;
+
+	D_ASSERT(tgt_cnt > 0);
+	D_ASSERT(tgt_ids != NULL);
+
+	D_ALLOC_PTR(eta);
+	if (eta == NULL)
+		return NULL;
+
+	D_ALLOC_ARRAY(eta->eta_ranks, tgt_cnt);
+	if (eta->eta_ranks == NULL)
+		goto free;
+
+	D_ALLOC_ARRAY(eta->eta_tgts, tgt_cnt);
+	if (eta->eta_tgts == NULL)
+		goto free;
+
+	uuid_copy(eta->eta_pool_id, pool_id);
+	eta->eta_pl_rank = pl_rank;
+	eta->eta_nr = tgt_cnt;
+	rank = dss_self_rank();
+	for (i = 0; i < eta->eta_nr; i++) {
+		eta->eta_ranks[i] = rank;
+		eta->eta_tgts[i] = tgt_ids[i];
+	}
+
+	return eta;
+free:
+	free_exclude_targets_arg(eta);
+	return NULL;
+}
+
+static void
+exclude_targets_ult(void *arg)
+{
+	struct exclude_targets_arg	*eta = arg;
+	struct d_tgt_list		 tgt_list;
+	d_rank_list_t			 svc;
+	int				 rc;
+
+	svc.rl_ranks = &eta->eta_pl_rank;
+	svc.rl_nr = 1;
+
+	tgt_list.tl_nr = eta->eta_nr;
+	tgt_list.tl_ranks = eta->eta_ranks;
+	tgt_list.tl_tgts = eta->eta_tgts;
+
+	rc = dsc_pool_tgt_exclude(eta->eta_pool_id, NULL /* grp */, &svc,
+				  &tgt_list);
+	if (rc)
+		D_ERROR(DF_UUID": Exclude targets failed. %d\n",
+			DP_UUID(eta->eta_pool_id), rc);
+
+	free_exclude_targets_arg(eta);
+}
+
+/*
+ * The NVMe faulty reaction is called from bio_nvme_poll() which is on
+ * progress (hardware poll) ULT, and it will call into client stack to
+ * exclude pool targets, blocking calls could be made in this code path,
+ * so we have to perform the faulty reactions asynchronously in a new ULT
+ * to avoid blocking the hardware poll.
+ */
+static int
+exclude_pool_targets(uuid_t pool_id, int *tgt_ids, int tgt_cnt,
+		     d_rank_t pl_rank)
+{
+	struct exclude_targets_arg	*eta;
+	int				 rc;
+
+	eta = alloc_exclude_targets_arg(pool_id, tgt_ids, tgt_cnt, pl_rank);
+	if (eta == NULL)
+		return -DER_NOMEM;
+
+	rc = dss_ult_create(exclude_targets_ult, eta, DSS_ULT_MISC,
+			    DSS_TGT_SELF, 0, NULL);
+	if (rc) {
+		D_ERROR(DF_UUID": Failed to start excluding ULT. %d\n",
+			DP_UUID(pool_id), rc);
+		free_exclude_targets_arg(eta);
+	}
+
+	return rc;
+}
+
+static int
+nvme_faulty_reaction(int *tgt_ids, int tgt_cnt)
+{
+	struct smd_pool_info	*pool_info, *tmp;
+	d_list_t		 pool_list;
+	d_rank_t		 pl_rank;
+	int			 pool_cnt, ret, rc;
+
+	D_ASSERT(tgt_cnt > 0);
+	D_ASSERT(tgt_ids != NULL);
+
+	D_INIT_LIST_HEAD(&pool_list);
+	rc = smd_pool_list(&pool_list, &pool_cnt);
```

Comment:
Instead of listing smd_pool_info at here, would it be clearer to list within BIO, and call the reaction callback for each pool?

Reply:
probably, actually I weighed these two options while coding.

I finally opted current way because I think it's better to make BIO pass 'local targets' to reaction interface, and rely on reaction callback implemented in pool module to convert the local targets to pool targets. We can revise it if we found that's not a good choice in the future.

Reply:
OK, let's keep it as now, but I still think traversing all pools is not functionality of the current pool module.